### PR TITLE
Tests for handle_action

### PR DESCRIPTION
**Pull request by IDragonfire**
CREATE_ACCOUNT tests are missing and waiting on #23
Does the email sending working?
I had encoding problems on my first mock test here:
`base64.b64encode(str(uid))`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,4 +92,15 @@ Can be combined !, e.g. `{command: social, teaminvite: <...>, friends: <..>}`
 
 ##### Misc
 
-* `{command: ask_session}`: response with an welcome commannd and an valid session (can be delyed)
\ No newline at end of file
+* `{command: ask_session}`: response with an welcome commannd and an valid session (can be delyed)
+
+## Stream
+
+It is possible to send data directly as Stream to the server. The response is normally a JSON Packet.
```
Comment (Sheeo) [inline README.md]: We should add that this section of the API is deprecated

File: tests/unit_tests/test_lobbyconnection.py [hunk 2]
```diff
@@ -255,3 +256,230 @@ def test_avatar_select_no_avatar(mock_query, fa_server_thread):
...
+# handle action
+def test_handle_action_ping(fa_server_thread):
+    fa_server_thread.sendReply = mock.Mock()
+    fa_server_thread.handleAction('PING', mock.Mock())
+    fa_server_thread.sendReply.assert_called_once_with('PONG')
+
+def test_handle_action_pong(fa_server_thread):
+    assert fa_server_thread.ponged == False
+    fa_server_thread.handleAction('PONG', mock.Mock())
+    assert fa_server_thread.ponged == True
```
Comment (Sheeo) [inline tests/unit_tests/test_lobbyconnection.py]: I think this is a bad test.

`lobbyconnection.ponged` should for one, not exist (We should use the `SO_KEEPALIVE` option) and if it should exist it should be a timestamp.

We should either have a TODO here or redo the implementation now (I don't think it's high priority)

**Comment by IDragonfire** [comment]
I am only a tester, would should I do?

File: tests/unit_tests/test_lobbyconnection.py [hunk 3]
```diff
@@ -255,3 +256,230 @@ def test_avatar_select_no_avatar(mock_query, fa_server_thread):
...
+    fa_server_thread.handleAction('FA_CLOSED', mock.Mock())
+    fa_server_thread.player.setAction.assert_called_once_with('NOTHING')
+    fa_server_thread.player.gameThread.abort.assert_called_once_with()
+
+def test_handle_action_unknownData(fa_server_thread):
+    fa_server_thread.receiveJSON = mock.Mock()
+    stream = mock.Mock()
+    fa_server_thread.handleAction('CrazyThing', stream)
+    fa_server_thread.receiveJSON.assert_called_once_with(\
+        'CrazyThing', stream)
```
Comment (Sheeo) [inline tests/unit_tests/test_lobbyconnection.py]: What exactly does this test?

File: tests/unit_tests/test_lobbyconnection.py [hunk 4]
```diff
@@ -255,3 +256,230 @@ def test_avatar_select_no_avatar(mock_query, fa_server_thread):
...
+                        'author': '', 'ui_only': '', 'version': '',
+                        'small': '', 'big': ''})
+    stream.readQString.side_effect = ['', '', zipMap, infos, 0, mock.Mock()]
+
+    # fake no db entry exists
+    mock_query.return_value.size.return_value = 0
+
+    fa_server_thread.handleAction('UPLOAD_MOD', stream)
+    fa_server_thread.sendJSON.assert_called_once_with( \
+        dict(command="notice", style="info", text="Mod correctly uploaded."))
```
Comment (Sheeo) [inline tests/unit_tests/test_lobbyconnection.py]: This test seems to assert that you may correctly upload a mod that has an empty name, no uid, no description etc...

Is that right?

File: tests/unit_tests/test_lobbyconnection.py [hunk 5]
```diff
@@ -255,3 +256,230 @@ def test_avatar_select_no_avatar(mock_query, fa_server_thread):
...
+                        'version': ''})
+    stream.readQString.side_effect = ['', '', zipMap, infos, 0, mock.Mock()]
+
+    # fake no db entry exists
+    mock_query.return_value.size.return_value = 0
+
+
+    fa_server_thread.handleAction('UPLOAD_MAP', stream)
+    fa_server_thread.sendJSON.assert_called_once_with( \
+        dict(command="notice", style="info", text="Map correctly uploaded."))
```
Comment (Sheeo) [inline tests/unit_tests/test_lobbyconnection.py]: Ditto

File: tests/unit_tests/test_lobbyconnection.py (hunk 4 above)
Comment (IDragonfire) [inline tests/unit_tests/test_lobbyconnection.py]: Correct... 
We have no checks... 
Only existing...

File: tests/unit_tests/test_lobbyconnection.py (hunk 3 above)
Comment (IDragonfire) [inline tests/unit_tests/test_lobbyconnection.py]: What happen if we receive invalid data.

Comment (IDragonfire) [inline tests/unit_tests/test_lobbyconnection.py]: That raise no exception... 
We store that...

File: tests/unit_tests/test_lobbyconnection.py [hunk 6]
```diff
@@ -255,3 +256,230 @@ def test_avatar_select_no_avatar(mock_query, fa_server_thread):
...
+             'map_type': '', 'battle_type': '', 'map_size': {'0': '', '1': ''},
+             'version': ''}
+
+    error_messages = {'name': 'No map name provided.',
+                      'description': 'No map description provided.',
+                      'max_players': 'No max players provided.',
+                      'map_type': 'No map type provided.',
+                      'battle_type': 'No battle type provided.',
+                      'map_size': 'No map size provided.',
+                      'version': 'No version provided.'}
```
Comment (Sheeo) [inline tests/unit_tests/test_lobbyconnection.py]: This only tests whether the message contains they keys, we should also verify each of them individually.

Comment (IDragonfire) [inline tests/unit_tests/test_lobbyconnection.py]: From my view we should keep that and switch to Web in the future....

Comment (Sheeo) [inline tests/unit_tests/test_lobbyconnection.py]: We will be doing that.

File: tests/unit_tests/test_lobbyconnection.py (hunk 4 above)
Comment (IDragonfire) [inline tests/unit_tests/test_lobbyconnection.py]: Maybe we have checks in the database?

**Comment by Sheeo** [comment]
`I had encoding problems on my first mock test here:
base64.b64encode(str(uid))` 

What kind of problems?

**Comment by IDragonfire** [comment]
Exception:
Str does not support the buffer interface....

**Comment by Sheeo** [comment]
Ah, that's a python3 thing. We'd need to do `base64.b64encode(str(uid).encode())`, but that's silly.

This is going away with @ChrisKitching's patch anyway, so it's fine.

File: tests/unit_tests/test_lobbyconnection.py (hunk 3 above)
Comment (IDragonfire) [inline tests/unit_tests/test_lobbyconnection.py]: THIS IS IMPORTANT:
over this way we can execute commands over the Stream API:
https://github.com/FAForever/server/blob/develop/src/lobbyconnection.py#L2413

**Comment by JeroenDeDauw** [comment]
This contains a number of serious http://c2.com/cgi/wiki?ArrangeActAssert violations. Uncountably this additional of tests is great, though don't forget one can have tech debt in tests themselves (and they can induce rigidity).

**Comment by Sheeo** [comment]
Yes, I'm aware of that, some of our other tests violate the rule too.

I was never a fan of strict rules though, sometimes introducing an extra test just for an additional assert is overkill.

**Comment by JeroenDeDauw** [comment]
I am also not a fan of strict rules. Reality is too complex for a single rule to make sense all of the time.

Violation of AAA in places where it actually is bad seems to be a common pattern here though, hence me pointing it out once again.

**Comment by Sheeo** [comment]
Could you mention a specific bad test and suggest improvement?

**Comment by JeroenDeDauw** [comment]
`test_fa_state`

**Comment by IDragonfire** [comment]
What should I improve and how?

**Comment by IDragonfire** [comment]
PS.: the problem from my view are not the tests...  More the units that are not decoupled will...

**Pull request merged**